gcp/apis/compute/v1alpha1/network_types.go

```diff
@@ -0,0 +1,390 @@
+	Status NetworkStatus `json:"status,omitempty"`
+}
+
+// SetBindingPhase of this ReplicationGroup.
+func (n *Network) SetBindingPhase(p v1alpha1.BindingPhase) {
+	n.Status.SetBindingPhase(p)
+}
+
+// SetConditions of this ReplicationGroup.
+func (n *Network) SetConditions(c ...v1alpha1.Condition) {
+	n.Status.SetConditions(c...)
+}
+
+// GetBindingPhase of this ReplicationGroup.
+func (n *Network) GetBindingPhase() v1alpha1.BindingPhase {
+	return n.Status.GetBindingPhase()
+}
+
+// SetClaimReference of this ReplicationGroup.
+func (n *Network) SetClaimReference(r *corev1.ObjectReference) {
+	n.Spec.ClaimReference = r
+}
+
+// GetClaimReference of this ReplicationGroup.
+func (n *Network) GetClaimReference() *corev1.ObjectReference {
+	return n.Spec.ClaimReference
+}
+
+// SetClassReference of this ReplicationGroup.
+func (n *Network) SetClassReference(r *corev1.ObjectReference) {
+	n.Spec.ClassReference = r
+}
+
+// GetClassReference of this ReplicationGroup.
+func (n *Network) GetClassReference() *corev1.ObjectReference {
+	return n.Spec.ClassReference
+}
+
+// SetWriteConnectionSecretToReference of this ReplicationGroup.
+func (n *Network) SetWriteConnectionSecretToReference(r corev1.LocalObjectReference) {
+	n.Spec.WriteConnectionSecretToReference = r
+}
+
+// GetWriteConnectionSecretToReference of this ReplicationGroup.
+func (n *Network) GetWriteConnectionSecretToReference() corev1.LocalObjectReference {
+	return n.Spec.WriteConnectionSecretToReference
+}
+
+// GetReclaimPolicy of this ReplicationGroup.
+func (n *Network) GetReclaimPolicy() v1alpha1.ReclaimPolicy {
+	return n.Spec.ReclaimPolicy
+}
+
+// SetReclaimPolicy of this ReplicationGroup.
+func (n *Network) SetReclaimPolicy(p v1alpha1.ReclaimPolicy) {
+	n.Spec.ReclaimPolicy = p
+}
+
+// +kubebuilder:object:root=true
+
+// NetworkList contains a list of Network
+type NetworkList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Network `json:"items"`
+}
+
+// GCPNetworkSpec contains fields of googlecompute.Network object that are
+// configurable by the user, i.e. the ones that are not marked as [Output Only]
+// In the future, this can be generated automatically.
+type GCPNetworkSpec struct {
+	// IPv4Range: Deprecated in favor of subnet mode networks. The range of
+	// internal addresses that are legal on this network. This range is a
+	// CIDR specification, for example: 192.168.0.0/16. Provided by the
+	// client when the network is created.
+	IPv4Range string `json:"IPv4Range,omitempty"`
+
+	// AutoCreateSubnetworks: When set to true, the VPC network is created
+	// in "auto" mode. When set to false, the VPC network is created in
+	// "custom" mode.
+	//
+	// An auto mode VPC network starts with one subnet per region. Each
+	// subnet has a predetermined range as described in Auto mode VPC
+	// network IP ranges.
+	AutoCreateSubnetworks bool `json:"autoCreateSubnetworks,omitempty"`
```

Comment:
In Google's API, there are 3 scenarios for this option (tested via REST client):
- false -> Custom mode
- true -> Auto mode
- nil (no key/value in json) -> Legacy mode.

The problem is that compute api has that field marked as `omitempty` (in `vendor/google.golang.org/api/compute/v1/compute-gen.go:16837`), meaning that when json conversion happens `autoCreateSubnetworks` field is omitted if it's false. That leaves us with only 2 option: `true`(Auto mode) or `nil`(Legacy mode). Basically, there is no way to create a Network resource in Custom mode right now using compute/v1 golang API.

I have removed `omitempty` from its source code and here and things worked, but then you lose Legacy mode option. I think right way to handle this would've been using `*bool` but that's not the case.

While I was digging this, I found the following statement in https://godoc.org/google.golang.org/api#hdr-Versioning_and_Stability 
```
Due to the auto-generated nature of this collection of libraries, complete APIs or specific versions can appear or go away without notice.
As a result, you should always locally vendor any API(s) that your code relies upon. 
```

One option is to use our modified version of `google.golang.org/api/compute/v1/compute-gen.go` and have all 3 modes available via use of `*bool` or we can remove only `omitempty` and support only Auto and Custom modes. @negz what do you think? do you know how we can file a bug for this?

For newcomers, here is some info about these modes:
```
When an auto mode network is created, one subnet from each region is automatically created within it.
...
In addition to the automatically created subnets, you can add more subnets manually to auto mode networks, in regions you choose, using IP ranges outside of 10.128.0.0/9.

When a custom mode network is created, no subnets are automatically created.
...
You decide which subnets to create, in regions you choose, and using IP ranges you specify.

Legacy mode(no subnets can be added): https://cloud.google.com/vpc/docs/legacy
```

Reply:
Alright, apparently I wasn't the only one https://github.com/googleapis/google-api-go-client/issues/129 Problem solved.